### DISCUSSION TITLE
vue 自动登录 问题

### DISCUSSION THREAD

**Discussion by zqm971210** [zxuqian/frontend-questions#6] [求助]
## up主你好,最近我又遇到了一个问题,当我在做页面刷新后自动登录功能时,遇到了路由跳转的一些问题？
![image](https://user-images.githubusercontent.com/48125795/97249900-c27e5e80-183f-11eb-9562-c20e9dfb3df5.png)
我正在做的这个小项目 是用一个项目实现前台展示和后台管理的项目(第一次接触前端,写的虽然不是很好,但是真的很认真的在写)
这是项目的前台连接 http://www.ehbio.com/Ecker/dist/#/ 
这是项目的后台连接 http://www.ehbio.com/Ecker/dist/#/BackendManagement/ 账号是:admin密码是:ehbioadmin

当我登录后,我通过学习,用的是vuex进行状态管理的
![image](https://user-images.githubusercontent.com/48125795/97250453-ed1ce700-1840-11eb-8e91-b9969903d261.png)

这是我的 app.vue 文件 
![image](https://user-images.githubusercontent.com/48125795/97250528-1fc6df80-1841-11eb-896b-84dcdad01778.png)

这是我进入后台管理的 vue文件
![image](https://user-images.githubusercontent.com/48125795/97250541-281f1a80-1841-11eb-8641-829f8329b222.png)

因为我在 vuex  actions 中 的 自动登录函数把 router.push 的路径写死了 所以我在等人后进入其他页面刷新后 进入的不是我刷新时的页面,而是我写死的页面.  
我知道是我把路由写死了导致的问题,但是我不知道该如何解决.
就是当我在一个页面刷新后,刷新之后还是我刷新时的页面.
我想过通过 router.path 将路径传递给 自动登录函数 ,但是我通过  router.path  获取路径的时候得到的不是我想要的结果 

![image](https://user-images.githubusercontent.com/48125795/97251035-1be78d00-1842-11eb-9d7d-62934aaf062d.png)
我是在进入每一个后台子页面的时候访问 vuex中的 token 查看是否有数据
在登陆的时候将token储存到vuex中,这里也是写死的跳转路由(但是这里是我想要的效果)
![image](https://user-images.githubusercontent.com/48125795/97251347-a7f9b480-1842-11eb-8632-7a600d257d25.png)

目前我实在进入后台页面的 beforeEnter 方法中进行自动登入的,我也尝试在 后台vue文件中的 create方法中进行自动登入 ,并且获取 路由 但是都达不到我想要的效果,
请问我改怎么解决这个问题呢? 
(我尝试了一些方法都没有解决 ,才来问的up主)

**Comment by zxuqian** [comment]
刷新之后就等于重新加载了 Vue 程序，Vuex 里的数据也会全部消失，所以 `store.state.userInfo.token` 在刷新之后是访问不到的，应该使用 localStorage。

**Comment by nonzzz** [comment]
能提供一下vuex的代码吗。如果 你要用store.state.userInfo.token 进行操作。应该给vuex 增加个插件。vuex-persist。然后在 store目录下/index.js导入插件就行

**Comment by zqm971210** [comment]
好的 谢谢

**Comment by zqm971210** [comment]
> 能提供一下vuex的代码吗。如果 你要用store.state.userInfo.token 进行操作。应该给vuex 增加个插件。vuex-persist。然后在 store目录下/index.js导入插件就行

不好意思 这么久才回复你 这几天有点忙 忘记了 问题 已经解决了 换了一种方式 
`import Vue from 'vue'
import Vuex from 'vuex'
import axios from "axios";
import {Notification} from 'element-ui'
import router from "../router";

Vue.use(Vuex)

export default new Vuex.Store({
  state: {
    userInfo: {},
    projectTitle: false
  },
  getters: {
    isNotLogin(state) {
      return state.userInfo.token
    },
    getHeaderInfo(state) {
      return state.userInfo.token
    },
  },
  mutations: {
    // 保存 注册登录信息
    saveUserInfo(state, userInfo) {
      state.userInfo = userInfo
    },
    // 清空用户登录状态
    clearUserInfo(state) {
      state.userInfo = {}
    },
    // 修改项目名
    changeProjectTitle(state, title) {
      state.projectTitle = title
    }
  },
  actions: {
    // 登录
    login({commit}, formData) {
      axios({
        url: '/change_information/ehbio_login/',
        method: 'post',
        data: formData
      }).then(res => {
        if (res.data.msg === "no user") {
          Notification({
            title: '登录失败!',
            message: `用户名不存在`,
            type: 'warning'
          })
        } else if (res.data.msg === "password error") {
          Notification({
            title: '登录失败!',
            message: `密码错误`,
            type: 'warning'
          })
        } else {
          commit('saveUserInfo', res.data)
          // 保存token
          localStorage.setItem('token', res.data.token)
          router.push({path: "/BackendManagement/IndexArticle"})
        }
      }).catch(err => {
        console.log(err)
      })
    },
    // 登出
    logout({commit}, token) {
      commit('clearUserInfo')
      localStorage.removeItem('token')
      axios({
        url: "/change_information/auto_logout/",
        method: "post",
        data: {token}
      })
    },
    //自动登录
    autoLogin({commit, dispatch}) {
      let token = localStorage.getItem('token')
      if (token) {
        axios({
          url: "/change_information/auto_login/",
          method: "post",
          data: {token}
        }).then((res) => {
          if (res.data.code) {
            commit('saveUserInfo', res.data)
            localStorage.setItem('token', res.data.token)
            router.push({path: "/BackendManagement/IndexArticle"})
          } else {
            Notification({
              title: '登录失败!',
              message: `请重新登录`,
              type: 'warning'
            })
          }
        })
      }
      dispatch('getHeaderInfo')
    },

    getHeaderInfo({commit}) {
      axios({
        url: '/change_information/get_lab_detail/',
        method: 'get',
      }).then((res) => {
        if (res.data.code) {
          document.title = res.data.data.name
          commit('changeProjectTitle', res.data.data.name)
        }
      }).catch(err => {
        console.log(err)
      });
    }
  },
  modules: {}
})
`这是我vuex的代码 写的不是很好
